### PR TITLE
mm: fix memory alignment

### PR DESCRIPTION
**Pull request by yamt**
## Testing
tested with https://github.com/apache/incubator-nuttx-apps/pull/1125
also this fixed a crash seen with examples/sotest on sim/macOS

### CODE REVIEW

File: mm/mm_heap/mm_memalign.c [hunk 1]
```diff
@@ -227,5 +230,7 @@ FAR void *mm_memalign(FAR struct mm_heap_s *heap, size_t alignment,
   kasan_unpoison((FAR void *)alignedchunk,
                  mm_malloc_size((FAR void *)alignedchunk));
 
-  return (FAR void *)alignedchunk;
+  ptr = (FAR void *)alignedchunk;
+  DEBUGASSERT(ptr == NULL || ((uintptr_t)ptr) % alignment == 0);
+  return ptr;
```
Comment (pkarashchenko) [inline mm/mm_heap/mm_memalign.c:235]: Maybe just
```suggestion
  DEBUGASSERT(alignedchunk % alignment == 0);
  return (FAR void *)alignedchunk;
```
to reduce number of lines in the PR.

File: mm/mm_heap/mm_memalign.c [hunk 2]
```diff
@@ -78,7 +79,9 @@ FAR void *mm_memalign(FAR struct mm_heap_s *heap, size_t alignment,
 
   if (alignment <= MM_MIN_CHUNK)
     {
-      return mm_malloc(heap, size);
+      ptr = mm_malloc(heap, size);
+      DEBUGASSERT(ptr == NULL || ((uintptr_t)ptr) % alignment == 0);
+      return ptr;
```
Comment (pkarashchenko) [inline mm/mm_heap/mm_memalign.c:84]:
```suggestion
      alignedchunk = (size_t)mm_malloc(heap, size);
      DEBUGASSERT(alignedchunk % alignment == 0);
      return (FAR void *)alignedchunk;
```

**Review by pkarashchenko — COMMENTED** [review]
All comments a minor

File: mm/mm_heap/mm_initialize.c [hunk 3]
```diff
@@ -93,12 +93,17 @@ void mm_addregion(FAR struct mm_heap_s *heap, FAR void *heapstart,
...
+   * Note: (uintptr_t)node + SIZEOF_MM_ALLOCNODE is what's actually
+   * returned to the malloc user, which should have natural alignment.
+   * (that is, in this implementation, MM_MIN_CHUNK-alignment.)
    */
 
-  heapbase = MM_ALIGN_UP((uintptr_t)heapstart);
-  heapend  = MM_ALIGN_DOWN((uintptr_t)heapstart + (uintptr_t)heapsize);
+  heapbase = MM_ALIGN_UP((uintptr_t)heapstart + SIZEOF_MM_ALLOCNODE)
+             - SIZEOF_MM_ALLOCNODE;
+  heapend  = MM_ALIGN_DOWN((uintptr_t)heapstart + (uintptr_t)heapsize)
```
Comment (xiaoxiang781216) [inline mm/mm_heap/mm_initialize.c:105]: Do we still need this patch? https://github.com/apache/incubator-nuttx/pull/5906 should already fix this issue.

Comment (yamt) [reply mm/mm_heap/mm_initialize.c:105]: yes.
i'm sure i tested recent enough version containing https://github.com/apache/incubator-nuttx/pull/5906

File: mm/mm_heap/mm_memalign.c (hunk 1 above)
Comment (yamt) [reply mm/mm_heap/mm_memalign.c:235]: i'm not sure what's the point of the suggested change.
is it that important to save a line? why?

**Comment by yamt** [comment]
i'm not sure why the ci is complaining on up_x11eventloop.c.

File: mm/mm_heap/mm_memalign.c (hunk 1 above)
Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:235]: 1. `0 % alignment  == 0` so we can eliminate `ptr == NULL` check
2. we can reuse existing variables and not introduce a new one.

Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:235]: anyway I wrote that all suggestions are optional and we can keep the code like it is now.

**Comment by pkarashchenko** [comment]
> i'm not sure why the ci is complaining on up_x11eventloop.c.

Try just to rebase the changes

**Comment by yamt** [comment]
> > i'm not sure why the ci is complaining on up_x11eventloop.c.
> 
> Try just to rebase the changes

why? is it a known problem already fixed?

File: mm/mm_heap/mm_memalign.c (hunk 1 above)
Comment (yamt) [reply mm/mm_heap/mm_memalign.c:235]: i prefer not to do it because bitwise operations on NULL is evil.

**Comment by pkarashchenko** [comment]
> > > i'm not sure why the ci is complaining on up_x11eventloop.c.
> > 
> > 
> > Try just to rebase the changes
> 
> why? is it a known problem already fixed?

https://github.com/apache/incubator-nuttx/pull/5988

**Comment by pkarashchenko** [comment]
But mainline seems to be broken now: https://github.com/apache/incubator-nuttx/pull/5937#issuecomment-1091152830
So probably need to hold on

File: mm/mm_heap/mm_memalign.c (hunk 1 above)
Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:235]: `alignedchunk` -- is `size_t` so not NULL based arithmetic operations

Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:235]: We can even update `alignedchunk` from `size_t` to `uintptr_t` to make it even more clear.

**Comment by yamt** [comment]
> > > > i'm not sure why the ci is complaining on up_x11eventloop.c.
> > > 
> > > 
> > > Try just to rebase the changes
> > 
> > 
> > why? is it a known problem already fixed?
> 
> #5988

thank you.
i was expecting that the ci only runs nxstyle on files modified in a PR.

**Comment by pkarashchenko** [comment]
> > > > > i'm not sure why the ci is complaining on up_x11eventloop.c.
> > > > 
> > > > 
> > > > Try just to rebase the changes
> > > 
> > > 
> > > why? is it a known problem already fixed?
> > 
> > 
> > #5988
> 
> thank you. i was expecting that the ci only runs nxstyle on files modified in a PR.

Yes. This should be true, but sometimes I see the issues like this. We can try just to restart the checks

**Comment by pkarashchenko** [comment]
What is actually does is 
```
6e0692c018 Merge fa714e92c5d53895f364fa920aaadb6028f0d659 into 53fef8d9c4457f3b4099997d4cbf894bbdb7146c
53fef8d9c4 arch/risc-v: Replace riscv_fault with riscv_exception
db012687f9 arch/sim: support simulator keyboard devices
b8a9d7da19 esp32: Add board support to TWAI/CANBus
[10](https://github.com/apache/incubator-nuttx/runs/5856746112?check_suite_focus=true#step:3:10)90e1a8ea xtensa/esp32: Add support to TWAI/CANBus controller
dfbf06e250 fix uinput rpmsg logic error
fa714e92c5 mm_initialize.c: Fix alignment
5d2c6b68[13](https://github.com/apache/incubator-nuttx/runs/5856746112?check_suite_focus=true#step:3:13) mm_malloc.c: Add assertions about alignment
ea1dc8595a mm_memalign.c: Add assertions about alignment
../nuttx/tools/checkpatch.sh -g ff0470d28a0875454f59fd6c87da0907d68ff6df..HEAD
```

**Comment by pkarashchenko** [comment]
https://github.com/apache/incubator-nuttx/pull/6002 -- here is a change to restore a CI on mainline

File: mm/mm_heap/mm_memalign.c (hunk 2 above)
Comment (yamt) [reply mm/mm_heap/mm_memalign.c:84]: i know (void *)0 is guaranteed to be NULL.
otoh, i'm not sure if (size_t)NULL is guaranteed to be 0.
i feel it's simpler to check NULL.

File: mm/mm_heap/mm_memalign.c (hunk 1 above)
Comment (yamt) [reply mm/mm_heap/mm_memalign.c:235]: i applied this one as it can't be NULL here.

File: mm/mm_heap/mm_initialize.c (hunk 3 above)
Comment (xiaoxiang781216) [reply mm/mm_heap/mm_initialize.c:105]: It's strange that the original code:

1. Align heapbase and heapend to MM_MIN_CHUNK
2. #5906 ensure two preallocate node with MM_MIN_CHUNK

So, the rest of free slot should multiple of MM_MIN_CHUNK.

File: mm/mm_heap/mm_memalign.c (hunk 2 above)
Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:84]: I just want to keep code consistent. We have code at line 112
```
  rawchunk = (size_t)mm_malloc(heap, allocsize);
  if (rawchunk == 0)
    {
      return NULL;
    }
```
so let's keep the same style if possible?

Comment (yamt) [reply mm/mm_heap/mm_memalign.c:84]: do you mean, within this file? or mm/? or the whole repo?

Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:84]: The ultimate goad is a whole repo of course :)
But let's focus on `mm_memalign` implementation only here.

Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:85]: Oah....I didn't expect that my comments will lead to adding even more local variables :)
I expected changes like
```
uintptr_t rawchunk;
...
rawchunk = (uintptr_t)mm_malloc(heap, allocsize);
if (rawchunk == (uintptr_t)NULL)
  {
    return NULL;
  }
```
and the same here, but not what I see.

File: mm/mm_heap/mm_initialize.c (hunk 3 above)
Comment (CV-Bowen) [inline mm/mm_heap/mm_initialize.c:105]:
```suggestion
  heapend  = MM_ALIGN_DOWN((uintptr_t)heapstart + (uintptr_t)heapsize) - SIZEOF_MM_ALLOCNODE;
```
This is also a solution to make heapsize align with MM_MIN_CHUNK and it seems better.

File: mm/mm_heap/mm_memalign.c (hunk 2 above)
Comment (yamt) [reply mm/mm_heap/mm_memalign.c:85]: do you prefer dropping all changes on mm_malloc.c and mm_memalign.c from this PR?
they are not essential parts of this PR. i feel it isn't that constructive to keep discussing about/changing them.

Comment (pkarashchenko) [reply mm/mm_heap/mm_memalign.c:85]: At least your previous version was more easy to read. That is my subjective opinion of course.
Agree, let's close the discussion.

File: mm/mm_heap/mm_initialize.c (hunk 3 above)
Comment (yamt) [reply mm/mm_heap/mm_initialize.c:105]: @CV-Bowen
even after this change, (heapend - heapbase) % MM_MIN_CHUNK == 0.
there is no problem wrt the goal of #5906 as far as i understand.

Comment (yamt) [reply mm/mm_heap/mm_initialize.c:105]: i don't understand your suggestion. the original patch already has `- SIZEOF_MM_ALLOCNODE`.

Comment (CV-Bowen) [reply mm/mm_heap/mm_initialize.c:105]: @yamt Sorry, i missed some code. So we satisfy the goal of #5991 and #5906 at the same time. Look good to me.

Comment (CV-Bowen) [reply mm/mm_heap/mm_initialize.c:105]: @xiaoxiang781216 @yamt we can satisfy the goal of https://github.com/apache/incubator-nuttx/pull/5991 and https://github.com/apache/incubator-nuttx/pull/5906 at the same time. Yesterday i missed some code and came to a wrong conclusion.

Comment (xiaoxiang781216) [reply mm/mm_heap/mm_initialize.c:105]: But, look like #5906 isn't need anymore?

**Comment by xiaoxiang781216** [comment]
@yamt could you review #6011? "mm_initialize.c: Fix alignment" could be dropped from this PR.

File: mm/mm_heap/mm_memalign.c [hunk 4]
```diff
@@ -106,13 +110,13 @@ FAR void *mm_memalign(FAR struct mm_heap_s *heap, size_t alignment,
 
   /* Then malloc that size */
 
-  rawchunk = (size_t)mm_malloc(heap, allocsize);
-  if (rawchunk == 0)
+  rawchunk_ptr = mm_malloc(heap, allocsize);
```
Comment (pkarashchenko) [inline mm/mm_heap/mm_memalign.c:113]: I still think that introduction of two local variables `ptr` and `rawchunk_ptr` is redundant and we can reuse `(FAR void *)` cast of existing `uintptr_t`, but that is minor and more related to code readability than to functionality.

**Comment by yamt** [comment]
i turned this a draft because i prefer https://github.com/apache/incubator-nuttx/pull/6011

**Comment by xiaoxiang781216** [comment]
@yamt could you rebase this patch and fix the conflict?

**Comment by xiaoxiang781216** [comment]
@yamt this PR contain more ASSERT which is very useful. How about we merge it after the conflict get resolved?

**Comment by xiaoxiang781216** [comment]
@yamt let finish the rest work in https://github.com/apache/incubator-nuttx/pull/7495

**Pull request closed without merging**